<p style="font-size:30px; text-align:center; line-height:120%">
<br>
COMS W 4111-02, H02, V02<br>
Introduction to Databases<br>
Spring 2019 Take Home Midterm
</p>

# Exam Overview and Instructions

- Homework assignments and exams have point values. Final grade depends on total point value. The range of total possible points for a semester is 0 to 100 points (not including extra-credit).
    - The grade for this midterm is in the range 0 to 100.
    - __This midterm exam is worth 20 points__ for your final semester points total. So, divide your score on this exam by 5 to determine point contribution.


- Submission:
    - The exam is due on 27-Oct-2019 at 11:59PM. __You may not use grace days.__
    - Submission:
        - Submission format a copy of this Jupyter Notebook with your solutions entered into the code cells or Markdown cells for each question. If you embed images or diagrams, you may need to use a zip file to include the images in the notebook.
        - You submit your homework on CourseWorks under "Midterm Examination" assignment.
        - No other formats are allowed.
        - If the notebook format is incorrect or the notebook is corrupted, the grade is 0.
        - Submissions after the due date and time are not allowed. Submissions not received on time receive a grade of 0.
    - Respect for the individual is paramount. We will accommodate special circumstances, but we must be notified and discuss __in advance.__
    

- Exam Rules:
    - Please read and review the [Academic Integrity policy and guidelines](https://www.college.columbia.edu/academics/academicintegrity), including subsections and details. This material defines the rules for this exam regardless of your school.
    - No collaboration of any form is permitted. You may not share material of any form, including links to on-line information, __information from the preparatory recitation section,__ suggestions or ideas, etc.
    - You __MAY__ use material from office hours or recitations.
    - You MAY use any on-line information you find, but may not directly use code you find. You must cite any on-line sources in the comments Markdown cell for each questions.
    - You must privately send comments or questions to Professor Ferguson via email. If questions or comments demonstrate a need for clarification or correction, Prof. Ferguson will update this document and send an email notification.
    - You may not discuss the exam on Piazza. The CA or professor may post clarifying comment on Piazza.


- Completing the exam:
    - Environment:
        - You must install [iPython-SQL](https://github.com/catherinedevlin/ipython-sql). Lecture notebooks have included iPython-SQL since lecture 1. There are installation instructions on the iPython-SQL web site and in lecture notes. __Note:__ You must install into your Anaconda environment, and not other system Python environments.
        - You will need to use several databases/schemas for the homework. You will need to use lahman2019clean, which you had to install for HW2. You will also need [classicmodels](http://www.mysqltutorial.org/mysql-sample-database.aspx). You need to install both databases, if you have not. You MUST use the schema/database names lahman2019clean and classmodels.
        - You must have a user ID dbuser with password dbuserdbuser and use in any connections you make completing questions.
        - Section 2 tests the setup of your environment. You __MUST USE__  dbuser:dbuserdbuser@localhost/lahman2019. 
        - Your database __MUST HAVE__ a user ```dbuser``` with pasword ```dbuserdbuser```
    - Each question starts with an explanation of the structure of the answer, e.g. prose, diagram, SQL, etc.
    - Questions requiring SQL or code have empty text cells where you enter your statements. Some also contain sample answers to help you determine if your answer is correct. You must show the execution of your SQL in your submission. You may use LIMIT 10 to reduce the size of results.

You may include diagrams in text/Markdown cells when answering a question. You can include a diagram using an HTML tag of the form:<br><br>

```<img src="./filename.file_extension">```

<br>
The example below between the horizontal lines includes an image. You can click on the cell to see the Markdown code for including the image. <b>You will have to submit a zip file containing the notebook and you image if you include diagrams or images in your submission.</b>
<br><br>

<hr style="height:2px;">

- Double click in between lines to see the Markdown example for including a diagram.
<br><br>
<img src="./Columbia_University_logo.jpg">
<br><br>
<hr style="height:2px;">

<span style="font-size:24px; color:red; line-height:1.2;">Note:</span>

- <span style="color:red;">Not all questions with the same point value are equally hard. </span>


- <span style="color:red;">One five point question might be much easier than another five point question.</span>


- <span style="color:red;">A ten point question is not always exactly twice as hard as a five point question.</span>


- <span style="color:red;">Students sometimes complain about the discrepancy in difficulty and value.</span>


- <span style="color:red;">I am sensitive to the concern. I am happy to resolve this complaint by redoing the exam to make all five point questions as hard as the hardest five point question, and all ten point questions as hard as the hardest ten point question.</span>


- <span style="color:red;">Just let me know.</span>

# Environment Test

This section tests the environment. You must change the "userid:pw" to the correct user ID and password for your MySQL instance. Please change back to "userid:pw" before submitting your exam. Unless you have received an exception, you __MUST USE dbuser:dbuserdbuser.__

## SQL Magic Plugin

In [62]:
%load_ext sql
%sql mysql+pymysql://dbuser:dbuserdbuser@localhost
        
%sql select * from lahman2019clean.people where playerid='willite01'

The sql extension is already loaded. To reload it, use:
  %reload_ext sql
 * mysql+pymysql://dbuser:***@localhost
1 rows affected.


[('willite01', '1918', '8', '30', 'USA', 'CA', 'San Diego', '2002', '7', '5', 'USA', 'FL', 'Inverness', 'Ted', 'Williams', 'Theodore Samuel', 205, '75', 'L', 'R', '1939-04-20', '1960-09-28', 'willt103', 'willite01', datetime.datetime(1918, 8, 30, 0, 0), datetime.datetime(2002, 7, 5, 0, 0))]

In [63]:
%sql select * from classicmodels.customers where customerNumber=103

 * mysql+pymysql://dbuser:***@localhost
1 rows affected.


[(103, 'Atelier graphique', 'Schmitt', 'Carine ', '40.32.2555', '54, rue Royale', None, 'Nantes', None, '44000', 'France', 1370, Decimal('21000.00'))]

## Python Connection

In [3]:
import json
import pymysql
import logging

logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger()
logger.setLevel(logging.DEBUG)

midterm_conn = pymysql.connect(
    host="localhost",
    user="dbuser",
    password="dbuserdbuser",
    cursorclass=pymysql.cursors.DictCursor)

In [4]:
import logging


def run_q(sql, args=None, fetch=True, cur=None, conn=midterm_conn, commit=True):
    '''
    Helper function to run an SQL statement.

    This is a modification that better supports HW1. An RDBDataTable MUST have a connection specified by
    the connection information. This means that this implementation of run_q MUST NOT try to obtain
    a defailt connection.

    :param sql: SQL template with placeholders for parameters. Canno be NULL.
    :param args: Values to pass with statement. May be null.
    :param fetch: Execute a fetch and return data if TRUE.
    :param conn: The database connection to use. This cannot be NULL, unless a cursor is passed.
        DO NOT PASS CURSORS for HW1.
    :param cur: The cursor to use. This is wizard stuff. Do not worry about it for now.
        DO NOT PASS CURSORS for HW1.
    :param commit: This is wizard stuff. Do not worry about it.

    :return: A pair of the form (execute response, fetched data). There will only be fetched data if
        the fetch parameter is True. 'execute response' is the return from the connection.execute, which
        is typically the number of rows effected.
    '''

    cursor_created = False
    connection_created = False

    try:

        if conn is None:
            raise ValueError("In this implementation, conn cannot be None.")

        if cur is None:
            cursor_created = True
            cur = conn.cursor()

        if args is not None:
            log_message = cur.mogrify(sql, args)
        else:
            log_message = sql

        logger.debug("Executing SQL = " + log_message)

        res = cur.execute(sql, args)

        if fetch:
            data = cur.fetchall()
        else:
            data = None

        # Do not ask.
        if commit == True:
            conn.commit()

    except Exception as e:
        raise(e)

    return (res, data)

In [5]:
q = "select playerID, nameLast, nameFirst from lahman2019clean.people where nameLast=%s and birthCity=%s"
res,d = run_q(q, args=('Williams', 'San Diego'))

print("Data = ", json.dumps(d, indent=2))

DEBUG:root:Executing SQL = select playerID, nameLast, nameFirst from lahman2019clean.people where nameLast='Williams' and birthCity='San Diego'


Data =  [
  {
    "playerID": "willite01",
    "nameLast": "Williams",
    "nameFirst": "Ted"
  },
  {
    "playerID": "willitr01",
    "nameLast": "Williams",
    "nameFirst": "Trevor"
  }
]


# Written Questions

Each question is worth 5 points.

## Benefits of Database Management Systems

- Prior to database management systems (DBMS), user relied on application programs that directly access files to create, retrieve and update shared data.


- Give five benefits of using a database management system to provide data access for applications.


- One or two sentences is sufficient for each answer.


- Double click on the number to open the Markdown cell.

<u>Answer</u>

1.  **Elimination of data redundancy and ease of data sharing:** The data is stored in a central database and shared via abstractions with all applications that require data. The applications can be developed using the abstractions provided by DBMS and not have to read and write to files independently.

2. **Data Standardization/Data Integrity:** DBMS provides several tools to manage data integrity and maintain sanctity of the data. Since the database is managed centrally for many applications, the standards are applied once and used by many.

3.  **Data security and entitlements:** DBMS ensures only authorized users are able to access the data based on their entitlements. Applications can use this functionality to control access to the users on what can be read, updated or created. 

4.  **Concurrency Controls:** DBMS handles the requests from concurrent users effectively to ensure data consistency. This otherwise requires significant effort to code using files.

5.  **Provide backup and recoverability:** DBMS supports an architecture to ensure recoverability from hardware or software failures and high availability. It also provides tools to take regular backup and frequent checkpoints which could be used to restore data when there is a corruption or any system issue. These are not easy to manage using files at every application level.

## Relational Concepts

Briefly explain _Cartesian product, equijoin, natural join,_ and _theta join._


<u>Answer</u>

**Cartesian Product:**  
It is also referred to as Cross-Join, and returns all the rows from the tables listed in the query. Let's say if we have two tables, Table1 and Table2, each row in the first table is paired with all the rows in the second table. This happens when there is no relationship defined between the two tables. Consider Table1 with 10 rows and Table2 with 20 rows, the cartesian product will return 200 (10 * 20) rows. Example query:

SELECT * From BirthInfo, AstrologicalInfo

**Equi join:**  
An equi-join is a specific type of comparator-based join, that uses only equality comparisons in the join-predicate. Using other comparison operators (such as >) disqualifies a join as an equi-join. Example of an equi-join query:

SELECT *
FROM BirthInfo JOIN AstrologicalInfo
  ON BirthInfo.Month = AstrologicalInfo.Month;
  
**Natural Join:**  

A Natural join is a join operation that generates a join clause based on the "common" columns in the two tables being joined. Common columns are columns that have the same name in both tables. This is a special case of equi-join. Example of a Natural join query (Month is the common column):

SELECT *
FROM BirthInfo NATURAL JOIN AstrologicalInfo;

**Theta Join:**  
A join operation using a general join condition is called a theta join. Theta joins work for all comparison operators. Join columns need not be compared using the equality sign.  Example of Theta join query:  

SELECT *
    FROM BirthInfo JOIN AstrologicalInfo
    ON BirthInfo.Month < AstrologicalInfo.Month;


## Relational Algebra

Use the following tables when answering this question.

| Name    | Month |
|--------|-----------|
| Don | September |
| Meghna | June      |
| Aly    | January   |
| Ara    | September |
| Kirit  | May       |
|| __BirthInfo__ |



| Month | Sign |
|-----------|-----------|
| January   | Acquarius |
| September | Virgo     |
| June      | Gemini    |
| July      | Leo       |
|| __AstrologicalInfo__ |


Give the result of each of the relational algebra statements. You can provide your answer in text in the form:
<br><br>
column name, column name, ..., column name<br>
value, value, ..., value<br>
value, value, ..., value<br>
value, value, ..., value<br>

<span style="font-size:20px;">1. $\sigma_{Month="September"}(BirthInfo)\ \bowtie \ \pi_{Sign}(\sigma_{Month<"September"}))$</span>
<br><br>
<u>Answer</u>

Name, Month, Sign
'Don','September','Aquarius'
'Ara','September','Aquarius'
'Don','September','Gemini'
'Ara','September','Gemini'
'Don','September','Leo'
'Ara','September','Leo'


<span style="font-size:20px;">2. $\pi_{Name}(\sigma_{Month="December"}(AstrologicalSign))$</span>
<br><br>
<u>Answer</u>

No Data is returned from the above select operation. The project operation will fail with an error as we cannot project "Name" from AstrologicalInfo schema.

<span style="font-size:20px;">3. $BirthInfo \  \bowtie_{BirthInfo.Month=AstrologicalSign.Month} \ AtrologicalInfo $</span>
<br><br>
<u>Answer</u> 

Name, Month, Month, Sign  
'Don','September','September','Virgo'  
'Meghna','June','June','Gemini'  
'Aly','January','January','Aquarius'  
'Ara','September','September','Virgo'  


<span style="font-size:20px;">4. $\pi_{Month}(BirthInfo) \land   \pi_{Month}(AstrologicalInfo) $</span>
<br><br>
<u>Answer</u>

Month  
'September'  
'June'  
'January'  

<span style="font-size:20px;">5. Produce an SQL statement that is equivalent to $Student( \underline{UNI}, last_name, first_name, email)$</span>
<br><br>
<u>Answer</u>

CREATE TABLE Student  
(  
UNI varchar(12) PRIMARY KEY,  
lastname varchar(32),  
firstname varchar(32),  
email varchar(72)  
)  

## Relational Semantics

Provide a short (at most five sentences) answer to the following questions.

```1.``` __Codd's Twelve Rules__ define what it means for a DBMS to be relational. Briefly explain Rule 3, "Systematic treatment of null values."

<u>Answer</u>

Systematic treatment of null values: The DBMS must allow each field to remain null (or empty). Specifically, it must support a representation of "missing information and inapplicable information" that is systematic, distinct from all regular values (for example, "distinct from zero or any other number", in the case of numeric values), and independent of data type. If the data is not known or the data is missing or not applicable, it has to be represented by NULL. It is also implied that such representations must be manipulated by the DBMS in a systematic way.


```2.``` In a relational model, the domain for an attribute must be _atomic._ Briefly explain what this means. Given an example of a domain that is not atomic.

<u>Answer</u>

Atomic means data cannot be divided further. A column with atomic data can't have several values of the same type of data in the same column.

The attribute 'fullname' is not atomic because it can be further divded into lastname and firstname. Similarly 'address' is also not atomic as it can be furhter divided into 'street', 'city', 'state', 'zip' and 'country'

```3.``` Briefly explain _super key, candidate key,_ and _primary key._

<u>Answer</u>

*Consider the student table for the examples on keys:  
Student(UNI, FirstName, LastName, GraduationYear, SSN)*


**Super Key:**  
* The set of attributes which can uniquely identify a tuple is known as Super Key. 
* Examples of Super Keys: UNI, SSN, (UNI,FirstName), (SSN,FirstName) etc.


**Candidate Key:** 
* The minimal set of attributes which can uniquely identify a tuple is known as candidate key. 
* Adding zero or more attributes to candidate key generates super key. A candidate key is a super key but vice versa is not true. The value of Candidate Key is unique for every tuple.
* Examples of Candidate Keys: UNI and SSN

**Primary Key:**. 
* There can be more than one candidate key in a relation out of which one can be chosen as primary key. Primary key cannot have null values, and just like candidate key it identifies a tuple uniquely
* Example of Primary Key: UNI


```4.``` Briefly (two or three sentences) explain the following concepts: _domain constraint, table integrity constraints, referential integrity constraints._

<u>Answer</u>

**Domain Constraint:**  

Domain integrity ensures that all the data items in a column fall within a defined set of valid values. The data type of domain includes string, character, integer, time, date, currency, etc. The value of the attribute must be available in the corresponding domain.

**Table Integrity Constraints:**

Table/entity integrity constraint is an integrity rule which states that every table must have a primary key and that the column or columns chosen to be the primary key should be unique and not null. This is because the primary key value is used to identify individual rows in relation and if the primary key has a null value, then we can't identify those rows. 

**Referential Integrity Constraints:**

A referential integrity constraint is defined as part of an association between two entity types and enforced via foreign keys. It is used to ensure that the values in a field of one table have matching values in a corresponding field of another table. Example - if a foreign key in Table 1 refers to the Primary Key of Table 2, then every value of the Foreign Key in Table 1 must be null or be available in Table 2.


```5.``` What are referential integrity _cascading deletes_ and _cascading updates?_

<u>Answer</u>

Referential integrity cascading deletes and cascading updates are foreign key constraints that tell DBMS to perform certain actions when a primary key field in a primary key-foreign key relationship is deleted or updated. By using cascading referential integrity constraints, we can define the actions to be taken when a user tries to delete or update a key to which existing foreign keys point.   
If a user tries delete statement(s) which will affect the rows in the foreign key table, then those rows will be deleted when the primary key record is deleted. Similarly, if an update statement affects rows in the foreign key table, then those rows will be updated with the value from the primary key record after it has been updated.


# SQL Data Manipulation Language Questions

##  Batter Performance for Red Sox in 1960 (5 points)

- This query requires the following columns from ```Lahman2019clean:```
    - ```people.playerid, people.nameLast, people.bats```
    - ```batting.playerid, batting.ab, batting.h, batting.bb, batting.hr, batting.teamid, batting.yearid, batting.2b, batting.3b, batting.HR```
    
    
- The formula for on-base percentage is (H + BB)/(H + AB). We will denote on-base percentage as ```OBP.```


- Batting average is H/AB. We will denote this as AVG.


- Slugging Percentage:
    - In the ```Batting``` table, ```H``` is total hits.
    - The table lists three types of hits ```2B`` is doubles, ```3B``` is triples and ```HR``` is homeruns.
    There is a fourth type of hit, _singles_ that contributes to total hits but is not in the table. We will call this ```1B```
    - The formulate for slugging percentage, which we will denote as SLG, is<br><br>
\begin{equation}
\frac{1B + 2*2B + 3*3B + 4*HR}{AB}
\end{equation}


- The following table summarizes batting performance for BOS in 1960 for the top ten hitters, ordered by SLG. Write and execute the SQL to produce the table.

In [27]:
%%sql
use lahman2019clean;

 * mysql+pymysql://dbuser:***@localhost
0 rows affected.


[]

<u>Your query and execution</u>

In [34]:
%%sql 
SELECT A.playerid, A.nameLast, A.bats, B.H, B.AB, 1B, 2B, 3B, HR, RBI,
		round((H/AB),3) AVG, round((H + BB)/(BB + AB),3) OBP, round(((1B + (2*2B) + (3*3B) + (4*HR)) / AB),3)  SLG
FROM people A, 
	(SELECT playerid, teamid, yearid, format(sum(H),0) H, format(sum(2B),0) 2B, format(sum(3B),0) 3B, 
     format(sum(HR),0) HR, format(sum(AB),0) AB, sum(BB) BB, format(sum(RBI),0) RBI,
    (sum(H) - sum(2B) - sum(3B)- sum(HR)) 1B 
	FROM batting
	GROUP BY playerid, teamid, yearid) B
WHERE A.playerid = B.playerid
and B.teamid = 'BOS'
and B.yearid = '1960'
order by SLG desc
limit 10

 * mysql+pymysql://dbuser:***@localhost
10 rows affected.


[('willite01', 'Williams', 'L', '98', '310', 54.0, '15', '0', '29', '72', 0.316, 0.449, 0.645),
 ('pagliji01', 'Pagliaroni', 'R', '19', '62', 10.0, '5', '2', '2', '9', 0.306, 0.427, 0.548),
 ('geigega01', 'Geiger', 'L', '74', '245', 49.0, '13', '3', '9', '33', 0.302, 0.362, 0.49),
 ('wertzvi01', 'Wertz', 'L', '125', '443', 84.0, '22', '0', '19', '103', 0.282, 0.338, 0.46),
 ('thomsbo01', 'Thomson', 'R', '30', '114', 21.0, '3', '1', '5', '20', 0.263, 0.328, 0.439),
 ('nixonru01', 'Nixon', 'L', '81', '272', 56.0, '17', '3', '5', '33', 0.298, 0.33, 0.438),
 ('fornimi01', 'Fornieles', 'R', '6', '15', 6.0, '0', '0', '0', '1', 0.4, 0.4, 0.4),
 ('malzofr01', 'Malzone', 'R', '161', '595', 115.0, '30', '2', '14', '79', 0.271, 0.312, 0.398),
 ('runnepe01', 'Runnels', 'L', '169', '528', 136.0, '29', '2', '2', '35', 0.32, 0.401, 0.394),
 ('tasbywi01', 'Tasby', 'R', '108', '385', 83.0, '17', '1', '7', '37', 0.281, 0.365, 0.384)]

<u>My Answer</u>

10 rows affected.


[('willite01', 'Williams', 'L', '98', '310', 54.0, '15', '0', '29', '72', 0.316, 0.449, 0.645),
 ('pagliji01', 'Pagliaroni', 'R', '19', '62', 10.0, '5', '2', '2', '9', 0.306, 0.427, 0.548),
 ('geigega01', 'Geiger', 'L', '74', '245', 49.0, '13', '3', '9', '33', 0.302, 0.362, 0.49),
 ('wertzvi01', 'Wertz', 'L', '125', '443', 84.0, '22', '0', '19', '103', 0.282, 0.338, 0.46),
 ('thomsbo01', 'Thomson', 'R', '30', '114', 21.0, '3', '1', '5', '20', 0.263, 0.328, 0.439),
 ('nixonru01', 'Nixon', 'L', '81', '272', 56.0, '17', '3', '5', '33', 0.298, 0.33, 0.438),
 ('fornimi01', 'Fornieles', 'R', '6', '15', 6.0, '0', '0', '0', '1', 0.4, 0.4, 0.4),
 ('malzofr01', 'Malzone', 'R', '161', '595', 115.0, '30', '2', '14', '79', 0.271, 0.312, 0.398),
 ('runnepe01', 'Runnels', 'L', '169', '528', 136.0, '29', '2', '2', '35', 0.32, 0.401, 0.394),
 ('tasbywi01', 'Tasby', 'R', '108', '385', 83.0, '17', '1', '7', '37', 0.281, 0.365, 0.384)]

## Set Membership (5 points)

- This query involves the lahman2019clean tables ```halloffame, people, appearances, pitching, managers.```


- Return the ```playerID, nameLast, nameFirst``` for every person that is in all of the tables.

<u>Your query and execution</u>

In [35]:
%%sql
SELECT playerID, nameFirst, nameLast FROM people a
WHERE 
    exists (select b.playerid from halloffame b where a.playerid = b.playerid) and
    exists (select c.playerid from appearances c where a.playerid = c.playerid) and
    exists (select d.playerid from pitching d where a.playerid = d.playerid) and
    exists (select e.playerid from managers e where a.playerid = e.playerid) 
ORDER BY playerid DESC
limit 10

# The above can also be written using joins as below. 
# I did not consider halloffame player only with inducted ='Y', since that is not asked in question

# SELECT playerID, nameFirst, nameLast FROM people
# WHERE playerid IN (
# 	SELECT a.playerid FROM halloffame a, pitching b, managers c, appearances d
# 	WHERE a.playerid = b.playerid
# 	AND a.playerid = c.playerid
# 	AND a.playerid = d.playerid)
# ORDER BY playerid DESC
# limit 10


 * mysql+pymysql://dbuser:***@localhost
10 rows affected.


[('youngcy01', 'Cy', 'Young'),
 ('wrighha01', 'Harry', 'Wright'),
 ('wrighge01', 'George', 'Wright'),
 ('willite01', 'Ted', 'Williams'),
 ('whitewi01', 'Will', 'White'),
 ('whitede01', 'Deacon', 'White'),
 ('wardjo01', 'John', 'Ward'),
 ('waltebu01', 'Bucky', 'Walters'),
 ('walshed01', 'Ed', 'Walsh'),
 ('wallabo01', 'Bobby', 'Wallace')]

<u>My Answer</u>

10 rows affected.


[('zimmech01', 'Chief', 'Zimmer'),
 ('yorkru01', 'Rudy', 'York'),
 ('wilsoji01', 'Jimmie', 'Wilson'),
 ('willsma01', 'Maury', 'Wills'),
 ('willima04', 'Matt', 'Williams'),
 ('willidi02', 'Dick', 'Williams'),
 ('westrwe01', 'Wes', 'Westrum'),
 ('weisswa01', 'Walt', 'Weiss'),
 ('wathajo01', 'John', 'Wathan'),
 ('walkeha01', 'Harry', 'Walker')]

## Complex Insert (10 points)


- Use ```classicmodels``` for this question.


- An order form typically looks something like:

| <img src="./order-form-103.jpg"> |
| :---: |
| __Example Order Form__ |


- For ```classicmodels``` the application user interface would ```POST``` of the form.

```
{
            "orderNumber": 10123,
            "orderDate": "2003-05-20",
            "requiredDate": "2003-05-29",
            "shippedDate": "2003-05-22",
            "status": "Shipped",
            "comments": null,
            "customerNumber": 103,
            "orderdetails": [
                {
                    "orderNumber": 10123,
                    "productCode": "S18_1589",
                    "quantityOrdered": 26,
                    "priceEach": "120.71",
                    "orderLineNumber": 2
                },
                {
                    "orderNumber": 10123,
                    "productCode": "S18_2870",
                    "quantityOrdered": 46,
                    "priceEach": "114.84",
                    "orderLineNumber": 3
                },
                {
                    "orderNumber": 10123,
                    "productCode": "S18_3685",
                    "quantityOrdered": 34,
                    "priceEach": "117.26",
                    "orderLineNumber": 4
                },
                {
                    "orderNumber": 10123,
                    "productCode": "S24_1628",
                    "quantityOrdered": 50,
                    "priceEach": "43.27",
                    "orderLineNumber": 1
                }
            ]
        }

```

- This data structure maps to two tables in ```classicmodels:``` ```orders``` and ```ordersdetails```


- Complete the implementation of the Python function below that takes a data structure (dict) of the form above and inserts that data into ```classicmodels.```


<u>Answer</u>


In [36]:
def create_order(order_info):
    """
    Creates (Inserts) the data associated with an order. The order information goes into orders table and each
    and line item/order detail item goes into the ordersdetails table.
    :param order_info: A dictionary. There are top-level elements for the order. There is an orderdetails element
        that is a list of dictionary for the orderdetails elements.
    :param cnx: The database connection to use.
    :return: A tuple of the form (order_insert_count, orderdetals_insert_count), where the values are the number
        of rows inserted into each table.
    """
    try:
        dtls = list(order_info.keys())
        
        ## Create the templates for inserting records into order and orderdetails tables.
        order = {}
        order_details_list = order_info['orderdetails']
        
        for key in dtls:
            if key == 'orderdetails':
                continue
            order[key]=order_info[key]

        ## Set table names
        schema_name = 'classicmodels'
        order_table_name = 'orders'
        order_details_table_name = 'orderdetails'
        full_order_table_name = schema_name+'.'+order_table_name
        full_order_details_table_name = schema_name+'.'+order_details_table_name
        
        ## Check if the Order Number already exists. If exists, log a message that it already exists
        chk = check_order_number(full_order_table_name, order['orderNumber'])
        if chk > 0:
            print ('The Order Number already exists!! generating a new valid order number...')
            
        orderNumber = get_valid_order_number(full_order_table_name)
        
        print ('The requested Order Number: ', order['orderNumber'], '; Generated Order Number: ', orderNumber )
        order['orderNumber'] = orderNumber
        for i in range(len(order_details_list)):
            order_details_list[i]['orderNumber'] = orderNumber
        
        
        ## Insert into orders table
        order_insert_count = insert(full_order_table_name, order)
        
        ## Insert into orderdetails table
        orderdetals_insert_count =0
        for i in range(len(order_details_list)):
            res = insert(full_order_details_table_name, order_details_list[i])
            orderdetals_insert_count += res
        

        ## Commit at the end if the inserts are succesfull on both tables
        midterm_conn.commit()
        
    except Exception as e:
        print ("Exception Occured !!! ", e)
        midterm_conn.rollback()
        return (0, 0)
        
    return (order_insert_count, orderdetals_insert_count)

def create_insert(table_name, new_row):

    sql = "insert into " + table_name + " "

    cols = list(new_row.keys())
    cols = ",".join(cols)
    col_clause = "(" + cols + ") "

    args = list(new_row.values())

    s_stuff = ["%s"]*len(args)
    s_clause = ",".join(s_stuff)
    v_clause = " values(" + s_clause + ")"

    sql += " " + col_clause + " " + v_clause

    return sql, args

def insert(table_name, new_record):
    """

    :param new_record: A dictionary representing a row to add to the set of records.
    :return: None
    """
    # Get the list of columns.
    sql, args = create_insert(table_name, new_record)
    res, d = run_q(sql, args=args, commit = False)
    return res

def check_order_number(table, key_value ,field_list = None):
    """
    """

    template = {}
    template['orderNumber'] = key_value

    w_clause, args = template_to_where_clause(template)

    if field_list is None:
        field_list = ['*']

    sql = "SELECT " + ",".join(field_list) + " FROM " + table + " " + w_clause

    result, data = run_q(sql=sql, args=args)

    if result is not None :
        result = result
    else:
        result = None

    return result

def template_to_where_clause(template):
    """

    :param template: One of those weird templates
    :return: WHERE clause corresponding to the template.
    """

    if template is None or template == {}:
        result = ("", None)
    else:
        args = []
        terms = []

        for k,v in template.items():
            terms.append(" " + k + "=%s ")
            args.append(v)

        w_clause = "AND".join(terms)
        w_clause = " WHERE " + w_clause

        result = (w_clause, args)

    return result


def get_valid_order_number(table):
    """
    """

    sql = "SELECT max(orderNumber) max_order_number FROM " + table 

    result, data = run_q(sql=sql)

    if data is not None :
        ret = (int(data[0]['max_order_number']) +1)
    else:
        ret = None
    
    print(data)
        
    return ret


In [37]:
info = {
            "orderNumber": 10123,
            "orderDate": "2003-05-20",
            "requiredDate": "2003-05-29",
            "shippedDate": "2003-05-22",
            "status": "Shipped",
            "comments": None,
            "customerNumber": 103,
            "orderdetails": [
                {
                    "orderNumber": 10123,
                    "productCode": "S18_1589",
                    "quantityOrdered": 26,
                    "priceEach": "120.71",
                    "orderLineNumber": 2
                },
                {
                    "orderNumber": 10123,
                    "productCode": "S18_2870",
                    "quantityOrdered": 46,
                    "priceEach": "114.84",
                    "orderLineNumber": 3
                },
                {
                    "orderNumber": 10123,
                    "productCode": "S18_3685",
                    "quantityOrdered": 34,
                    "priceEach": "117.26",
                    "orderLineNumber": 4
                },
                {
                    "orderNumber": 10123,
                    "productCode": "S24_1628",
                    "quantityOrdered": 50,
                    "priceEach": "43.27",
                    "orderLineNumber": 1
                }
            ]
        }

##### Here is the test case to test inserts. The process should log a message saying the order already exists and go-ahead with generating a valid order number.  

In [38]:
records_order, records_orderdtls = create_order(info)
print ('\n Records inserted to Orders table: ',records_order)
print (' Records inserted to Order Details table: ',records_orderdtls)


DEBUG:root:Executing SQL = SELECT * FROM classicmodels.orders  WHERE  orderNumber=10123 
DEBUG:root:Executing SQL = SELECT max(orderNumber) max_order_number FROM classicmodels.orders
DEBUG:root:Executing SQL = insert into classicmodels.orders  (orderNumber,orderDate,requiredDate,shippedDate,status,comments,customerNumber)   values(10428,'2003-05-20','2003-05-29','2003-05-22','Shipped',NULL,103)
DEBUG:root:Executing SQL = insert into classicmodels.orderdetails  (orderNumber,productCode,quantityOrdered,priceEach,orderLineNumber)   values(10428,'S18_1589',26,'120.71',2)
DEBUG:root:Executing SQL = insert into classicmodels.orderdetails  (orderNumber,productCode,quantityOrdered,priceEach,orderLineNumber)   values(10428,'S18_2870',46,'114.84',3)
DEBUG:root:Executing SQL = insert into classicmodels.orderdetails  (orderNumber,productCode,quantityOrdered,priceEach,orderLineNumber)   values(10428,'S18_3685',34,'117.26',4)
DEBUG:root:Executing SQL = insert into classicmodels.orderdetails  (orderN

The Order Number already exists!! generating a new valid order number...
[{'max_order_number': 10427}]
The requested Order Number:  10123 ; Generated Order Number:  10428

 Records inserted to Orders table:  1
 Records inserted to Order Details table:  4


In [39]:
info = {
            "orderNumber": 10427232,
            "orderDate": "2003-05-20",
            "requiredDate": "2003-05-29",
            "shippedDate": "2003-05-22",
            "status": "Shipped",
            "comments": None,
            "customerNumber": 103,
            "orderdetails": [
                {
                    "orderNumber": 10427232,
                    "productCode": "S18_1589",
                    "quantityOrdered": 26,
                    "priceEach": "120.71",
                    "orderLineNumber": 2
                },
                {
                    "orderNumber": 10427232,
                    "productCode": "S18_2870",
                    "quantityOrdered": 46,
                    "priceEach": "114.84",
                    "orderLineNumber": 3
                },
                {
                    "orderNumber": 10427232,
                    "productCode": "S18_3685",
                    "quantityOrdered": 34,
                    "priceEach": "117.26",
                    "orderLineNumber": 4
                },
                {
                    "orderNumber": 10427232,
                    "productCode": "S24_1628",
                    "quantityOrdered": 50,
                    "priceEach": "43.27",
                    "orderLineNumber": 1
                }
            ]
        }

##### Here is the test case where Order number provided doesn't exist in table. The process should still go-ahead and generate a valid order number for creating the records. In the below example, 10427232 doesn't exist but still the process decided to generate a new order number. Essentially, the process assumes that the order number provided in the record is never correct.

In [40]:
records_order, records_orderdtls = create_order(info)
print ('\n Records inserted to Orders table: ',records_order)
print (' Records inserted to Order Details table: ',records_orderdtls)

DEBUG:root:Executing SQL = SELECT * FROM classicmodels.orders  WHERE  orderNumber=10427232 
DEBUG:root:Executing SQL = SELECT max(orderNumber) max_order_number FROM classicmodels.orders
DEBUG:root:Executing SQL = insert into classicmodels.orders  (orderNumber,orderDate,requiredDate,shippedDate,status,comments,customerNumber)   values(10429,'2003-05-20','2003-05-29','2003-05-22','Shipped',NULL,103)
DEBUG:root:Executing SQL = insert into classicmodels.orderdetails  (orderNumber,productCode,quantityOrdered,priceEach,orderLineNumber)   values(10429,'S18_1589',26,'120.71',2)
DEBUG:root:Executing SQL = insert into classicmodels.orderdetails  (orderNumber,productCode,quantityOrdered,priceEach,orderLineNumber)   values(10429,'S18_2870',46,'114.84',3)
DEBUG:root:Executing SQL = insert into classicmodels.orderdetails  (orderNumber,productCode,quantityOrdered,priceEach,orderLineNumber)   values(10429,'S18_3685',34,'117.26',4)
DEBUG:root:Executing SQL = insert into classicmodels.orderdetails  (ord

[{'max_order_number': 10428}]
The requested Order Number:  10427232 ; Generated Order Number:  10429

 Records inserted to Orders table:  1
 Records inserted to Order Details table:  4


## Complex Query/View $-$ Player Performance Statistics by Year (10 points)

- Use the ```lahman2019clean``` database/schema.


- Create performance summary views. Create five views:
    - ```batting_summary:``` yearID, teamID, AB, H, HR, RBI
    - ```appearances summary:``` yearID, teamID, G_all, GS
    - ```pitching summary:``` yearID, teamID, W, L, IPouts
    - ```fielding_summary``` yearID, teamID, PO, A, E, POS
    - ```annual_summary,``` which combines the views above.
    - ```career_summary,``` which contains the totals/summaries for the entire career.
    
    
- __NOTE:__ You will need to do aggregation on some of the views to get annual values.


- __Note:__ Your query must produce the correct results for any ```playerID.```


<u>Answer</u>


- ```batting_summary```

```
CREATE VIEW `lahman2019clean`.`batting_summary` 
AS SELECT playerid, teamid, yearid, sum(ab) ab, sum(h) h, sum(hr) hr, sum(rbi) rbi 
	FROM batting 
	GROUP BY playerid, teamid, yearid;
```

In [42]:
%sql select * from batting_summary where playerID='willite01'

 * mysql+pymysql://dbuser:***@localhost
19 rows affected.


[('willite01', 'BOS', '1939', 565.0, 185.0, 31.0, 145.0),
 ('willite01', 'BOS', '1940', 561.0, 193.0, 23.0, 113.0),
 ('willite01', 'BOS', '1941', 456.0, 185.0, 37.0, 120.0),
 ('willite01', 'BOS', '1942', 522.0, 186.0, 36.0, 137.0),
 ('willite01', 'BOS', '1946', 514.0, 176.0, 38.0, 123.0),
 ('willite01', 'BOS', '1947', 528.0, 181.0, 32.0, 114.0),
 ('willite01', 'BOS', '1948', 509.0, 188.0, 25.0, 127.0),
 ('willite01', 'BOS', '1949', 566.0, 194.0, 43.0, 159.0),
 ('willite01', 'BOS', '1950', 334.0, 106.0, 28.0, 97.0),
 ('willite01', 'BOS', '1951', 531.0, 169.0, 30.0, 126.0),
 ('willite01', 'BOS', '1952', 10.0, 4.0, 1.0, 3.0),
 ('willite01', 'BOS', '1953', 91.0, 37.0, 13.0, 34.0),
 ('willite01', 'BOS', '1954', 386.0, 133.0, 29.0, 89.0),
 ('willite01', 'BOS', '1955', 320.0, 114.0, 28.0, 83.0),
 ('willite01', 'BOS', '1956', 400.0, 138.0, 24.0, 82.0),
 ('willite01', 'BOS', '1957', 420.0, 163.0, 38.0, 87.0),
 ('willite01', 'BOS', '1958', 411.0, 135.0, 26.0, 85.0),
 ('willite01', 'BOS', '1959', 272.0, 69.0, 10.0, 43.0),
 ('willite01', 'BOS', '1960', 310.0, 98.0, 29.0, 72.0)]

- ```Pitching summary```

```
CREATE VIEW `lahman2019clean`.`pitching_summary` 
AS SELECT playerid, teamid, yearid, sum(w) w, sum(l) l, sum(g) g_p, sum(IPouts) IPouts 
	FROM pitching 
	GROUP BY playerid, teamid, yearid;```

In [43]:
%sql select * from pitching_summary where playerid='willite01';

 * mysql+pymysql://dbuser:***@localhost
1 rows affected.


[('willite01', 'BOS', '1940', 0.0, 0.0, 1.0, 6.0)]

- ```fielding summary```

```
CREATE VIEW `lahman2019clean`.`fielding_summary` 
AS SELECT playerid, teamid, yearid, sum(po) po, sum(a) a, sum(e) e, group_concat(pos) 
	FROM fielding 
	GROUP BY playerid, teamid, yearid;```


In [44]:
%sql select * from fielding_summary where playerid='willite01'

 * mysql+pymysql://dbuser:***@localhost
19 rows affected.


[('willite01', 'BOS', '1939', 318.0, 11.0, 19.0, 'OF'),
 ('willite01', 'BOS', '1940', 302.0, 15.0, 13.0, 'OF,P'),
 ('willite01', 'BOS', '1941', 262.0, 11.0, 11.0, 'OF'),
 ('willite01', 'BOS', '1942', 312.0, 15.0, 4.0, 'OF'),
 ('willite01', 'BOS', '1946', 325.0, 7.0, 10.0, 'OF'),
 ('willite01', 'BOS', '1947', 347.0, 10.0, 9.0, 'OF'),
 ('willite01', 'BOS', '1948', 289.0, 9.0, 5.0, 'OF'),
 ('willite01', 'BOS', '1949', 337.0, 12.0, 6.0, 'OF'),
 ('willite01', 'BOS', '1950', 165.0, 7.0, 8.0, 'OF'),
 ('willite01', 'BOS', '1951', 315.0, 12.0, 4.0, 'OF'),
 ('willite01', 'BOS', '1952', 4.0, 0.0, 0.0, 'OF'),
 ('willite01', 'BOS', '1953', 31.0, 1.0, 1.0, 'OF'),
 ('willite01', 'BOS', '1954', 213.0, 5.0, 4.0, 'OF'),
 ('willite01', 'BOS', '1955', 170.0, 5.0, 2.0, 'OF'),
 ('willite01', 'BOS', '1956', 174.0, 7.0, 5.0, 'OF'),
 ('willite01', 'BOS', '1957', 215.0, 2.0, 1.0, 'OF'),
 ('willite01', 'BOS', '1958', 154.0, 3.0, 7.0, 'OF'),
 ('willite01', 'BOS', '1959', 94.0, 4.0, 3.0, 'OF'),
 ('willite01', 'BOS', '1960', 131.0, 6.0, 1.0, 'OF')]

- ```appearances_summary```

```
CREATE VIEW `lahman2019clean`.`appearances_summary` 
AS SELECT playerid, teamid, yearid, G_all, GS
	FROM appearances;```

In [45]:
%sql select * from appearances_summary where playerid = 'willite01'

 * mysql+pymysql://dbuser:***@localhost
19 rows affected.


[('willite01', 'BOS', '1939', '149', '149'),
 ('willite01', 'BOS', '1940', '144', '143'),
 ('willite01', 'BOS', '1941', '143', '133'),
 ('willite01', 'BOS', '1942', '150', '150'),
 ('willite01', 'BOS', '1946', '150', '150'),
 ('willite01', 'BOS', '1947', '156', '156'),
 ('willite01', 'BOS', '1948', '137', '134'),
 ('willite01', 'BOS', '1949', '155', '155'),
 ('willite01', 'BOS', '1950', '89', '86'),
 ('willite01', 'BOS', '1951', '148', '147'),
 ('willite01', 'BOS', '1952', '6', '2'),
 ('willite01', 'BOS', '1953', '37', '26'),
 ('willite01', 'BOS', '1954', '117', '113'),
 ('willite01', 'BOS', '1955', '98', '93'),
 ('willite01', 'BOS', '1956', '136', '110'),
 ('willite01', 'BOS', '1957', '132', '125'),
 ('willite01', 'BOS', '1958', '129', '114'),
 ('willite01', 'BOS', '1959', '103', '75'),
 ('willite01', 'BOS', '1960', '113', '87')]

- ```annual_summary```

```
CREATE VIEW `lahman2019clean`.`annual_summary` 
AS SELECT a.playerid, a.teamid, a.yearid, a.G_all, a.GS, b.ab, b.h, b.hr, b.rbi, c.w, c.l, c.g_p, c.IPouts, d.po, d.a, d.e, d.`group_concat(pos)`        
FROM appearances_summary a 
    left join batting_summary b on a.playerid = b.playerid and a.teamid = b.teamid and a.yearid = b.yearid
    left join pitching_summary c on a.playerid = c.playerid and a.teamid = c.teamid and a.yearid = c.yearid
    left join fielding_summary d on a.playerid = d.playerid and a.teamid = d.teamid and a.yearid = d.yearid```

In [46]:
%sql select * from annual_summary where playerid='willite01'

 * mysql+pymysql://dbuser:***@localhost
19 rows affected.


[('willite01', 'BOS', '1939', '149', '149', 565.0, 185.0, 31.0, 145.0, None, None, None, None, 318.0, 11.0, 19.0, 'OF'),
 ('willite01', 'BOS', '1940', '144', '143', 561.0, 193.0, 23.0, 113.0, 0.0, 0.0, 1.0, 6.0, 302.0, 15.0, 13.0, 'OF,P'),
 ('willite01', 'BOS', '1941', '143', '133', 456.0, 185.0, 37.0, 120.0, None, None, None, None, 262.0, 11.0, 11.0, 'OF'),
 ('willite01', 'BOS', '1942', '150', '150', 522.0, 186.0, 36.0, 137.0, None, None, None, None, 312.0, 15.0, 4.0, 'OF'),
 ('willite01', 'BOS', '1946', '150', '150', 514.0, 176.0, 38.0, 123.0, None, None, None, None, 325.0, 7.0, 10.0, 'OF'),
 ('willite01', 'BOS', '1947', '156', '156', 528.0, 181.0, 32.0, 114.0, None, None, None, None, 347.0, 10.0, 9.0, 'OF'),
 ('willite01', 'BOS', '1948', '137', '134', 509.0, 188.0, 25.0, 127.0, None, None, None, None, 289.0, 9.0, 5.0, 'OF'),
 ('willite01', 'BOS', '1949', '155', '155', 566.0, 194.0, 43.0, 159.0, None, None, None, None, 337.0, 12.0, 6.0, 'OF'),
 ('willite01', 'BOS', '1950', '89', '86', 334.0, 106.0, 28.0, 97.0, None, None, None, None, 165.0, 7.0, 8.0, 'OF'),
 ('willite01', 'BOS', '1951', '148', '147', 531.0, 169.0, 30.0, 126.0, None, None, None, None, 315.0, 12.0, 4.0, 'OF'),
 ('willite01', 'BOS', '1952', '6', '2', 10.0, 4.0, 1.0, 3.0, None, None, None, None, 4.0, 0.0, 0.0, 'OF'),
 ('willite01', 'BOS', '1953', '37', '26', 91.0, 37.0, 13.0, 34.0, None, None, None, None, 31.0, 1.0, 1.0, 'OF'),
 ('willite01', 'BOS', '1954', '117', '113', 386.0, 133.0, 29.0, 89.0, None, None, None, None, 213.0, 5.0, 4.0, 'OF'),
 ('willite01', 'BOS', '1955', '98', '93', 320.0, 114.0, 28.0, 83.0, None, None, None, None, 170.0, 5.0, 2.0, 'OF'),
 ('willite01', 'BOS', '1956', '136', '110', 400.0, 138.0, 24.0, 82.0, None, None, None, None, 174.0, 7.0, 5.0, 'OF'),
 ('willite01', 'BOS', '1957', '132', '125', 420.0, 163.0, 38.0, 87.0, None, None, None, None, 215.0, 2.0, 1.0, 'OF'),
 ('willite01', 'BOS', '1958', '129', '114', 411.0, 135.0, 26.0, 85.0, None, None, None, None, 154.0, 3.0, 7.0, 'OF'),
 ('willite01', 'BOS', '1959', '103', '75', 272.0, 69.0, 10.0, 43.0, None, None, None, None, 94.0, 4.0, 3.0, 'OF'),
 ('willite01', 'BOS', '1960', '113', '87', 310.0, 98.0, 29.0, 72.0, None, None, None, None, 131.0, 6.0, 1.0, 'OF')]

- ```career_summary```

```
CREATE VIEW `lahman2019clean`.`career_summary` 
AS SELECT playerid, cast(sum(G_all) as unsigned) as g_all, cast(sum(GS) as unsigned) gs, 
		cast(sum(ab) as unsigned) ab, cast(sum(h) as unsigned) h, cast(sum(hr) as unsigned) hr, 
        cast(sum(rbi) as unsigned) rbi, cast(sum(w) as unsigned) w, cast(sum(l) as unsigned) l, 
        cast(sum(IPouts) as unsigned) IPouts, cast(sum(po) as unsigned) po, 
        cast(sum(a) as unsigned) a, cast(sum(e) as unsigned) e, 
        (select group_concat(distinct pos) from fielding b where a.playerId = b.playerID ) positions
FROM annual_summary a
GROUP BY playerID
```

In [47]:
%sql select * from career_summary limit 10;

 * mysql+pymysql://dbuser:***@localhost
10 rows affected.


[('aardsda01', 331, 0, 4, 32, 0, 0, 16, 18, 1011, 11, 29, 3, 'P'),
 ('aaronha01', 3298, 3173, 12364, 3771, 755, 2297, None, None, None, 7436, 429, 144, '1B,2B,3B,OF'),
 ('aaronto01', 437, 206, 944, 216, 13, 94, None, None, None, 1317, 113, 22, '1B,2B,3B,OF'),
 ('aasedo01', 448, 91, 5, 0, 0, 0, 66, 60, 3328, 67, 135, 13, 'P'),
 ('abadan01', 15, 4, 21, 2, 0, 0, None, None, None, 37, 1, 1, '1B,OF'),
 ('abadfe01', 363, 6, 9, 1, 0, 0, 8, 27, 953, 7, 37, 2, 'P'),
 ('abadijo01', 12, 0, 49, 11, 0, 5, None, None, None, 129, 3, 13, '1B'),
 ('abbated01', 857, 357, 3044, 772, 11, 324, None, None, None, 1873, 2368, 315, '2B,3B,OF,SS'),
 ('abbeybe01', 79, 0, 225, 38, 0, 17, 22, 40, 1704, 17, 134, 22, 'P'),
 ('abbeych01', 452, 0, 1756, 493, 19, 280, 0, 0, 6, 920, 92, 100, 'OF,P')]

## Update Statement (5 points)

- Make copies of the ```orders``` table in ```classicmodels.```


- The following statements will accomplish that.

In [48]:
%%sql
use classicmodels;
drop table if exists orders_copy;
create table orders_copy as select * from orders;

 * mysql+pymysql://dbuser:***@localhost
0 rows affected.
0 rows affected.
330 rows affected.


[]

- You can test of your copy worked by producing the same results as the following query.

In [49]:
%sql select * from orders_copy join orderdetails using(orderNumber) where orderNumber=10100;

 * mysql+pymysql://dbuser:***@localhost
4 rows affected.


[(10100, datetime.date(2003, 1, 6), datetime.date(2003, 1, 13), datetime.date(2003, 1, 10), 'Shipped', None, 363, 'S18_1749', 30, Decimal('136.00'), 3),
 (10100, datetime.date(2003, 1, 6), datetime.date(2003, 1, 13), datetime.date(2003, 1, 10), 'Shipped', None, 363, 'S18_2248', 50, Decimal('55.09'), 2),
 (10100, datetime.date(2003, 1, 6), datetime.date(2003, 1, 13), datetime.date(2003, 1, 10), 'Shipped', None, 363, 'S18_4409', 22, Decimal('75.46'), 4),
 (10100, datetime.date(2003, 1, 6), datetime.date(2003, 1, 13), datetime.date(2003, 1, 10), 'Shipped', None, 363, 'S24_3969', 49, Decimal('35.29'), 1)]

- Write a single ```UPDATE``` statement that sets the status of all orders for customers to 'EMBARGOED' if:
    - The customer's address is in Australia And
    - The order's status is not SHIPPED or CANCELLED.
    

- Before the update, run the following query. You should get results that match the example.

In [25]:
%%sql
select
	customers.customerNumber, customers.country, orders_copy.orderNumber, orders_copy.status from
    customers join orders_copy
    using (customerNumber)
    where country = 'Australia'
order by status;

 * mysql+pymysql://dbuser:***@localhost
19 rows affected.


[(471, 'Australia', 10415, 'Disputed'),
 (282, 'Australia', 10420, 'In Process'),
 (114, 'Australia', 10120, 'Shipped'),
 (114, 'Australia', 10125, 'Shipped'),
 (282, 'Australia', 10139, 'Shipped'),
 (276, 'Australia', 10148, 'Shipped'),
 (333, 'Australia', 10152, 'Shipped'),
 (276, 'Australia', 10169, 'Shipped'),
 (333, 'Australia', 10174, 'Shipped'),
 (471, 'Australia', 10193, 'Shipped'),
 (114, 'Australia', 10223, 'Shipped'),
 (471, 'Australia', 10265, 'Shipped'),
 (282, 'Australia', 10270, 'Shipped'),
 (114, 'Australia', 10342, 'Shipped'),
 (114, 'Australia', 10347, 'Shipped'),
 (282, 'Australia', 10361, 'Shipped'),
 (276, 'Australia', 10370, 'Shipped'),
 (333, 'Australia', 10374, 'Shipped'),
 (276, 'Australia', 10391, 'Shipped')]

<u>Answer</u> Your update statement

In [26]:
%%sql
UPDATE orders_copy SET status = 'EMBARGOED'
WHERE customerNumber in (select customerNumber from customers where country = 'Australia')
and upper(status) not in ('SHIPPED','CANCELLED')

 * mysql+pymysql://dbuser:***@localhost
2 rows affected.


[]

- After running your update, run the following query to produce the same output as the example.

In [27]:
%%sql
select
	customers.customerNumber, customers.country, orders_copy.orderNumber, orders_copy.status from
    customers join orders_copy
    using (customerNumber)
    where country = 'Australia'
order by status;

 * mysql+pymysql://dbuser:***@localhost
19 rows affected.


[(471, 'Australia', 10415, 'EMBARGOED'),
 (282, 'Australia', 10420, 'EMBARGOED'),
 (114, 'Australia', 10120, 'Shipped'),
 (114, 'Australia', 10125, 'Shipped'),
 (282, 'Australia', 10139, 'Shipped'),
 (276, 'Australia', 10148, 'Shipped'),
 (333, 'Australia', 10152, 'Shipped'),
 (276, 'Australia', 10169, 'Shipped'),
 (333, 'Australia', 10174, 'Shipped'),
 (471, 'Australia', 10193, 'Shipped'),
 (114, 'Australia', 10223, 'Shipped'),
 (471, 'Australia', 10265, 'Shipped'),
 (282, 'Australia', 10270, 'Shipped'),
 (114, 'Australia', 10342, 'Shipped'),
 (114, 'Australia', 10347, 'Shipped'),
 (282, 'Australia', 10361, 'Shipped'),
 (276, 'Australia', 10370, 'Shipped'),
 (333, 'Australia', 10374, 'Shipped'),
 (276, 'Australia', 10391, 'Shipped')]

# Data Modeling, Cleanup and Implementation

## Classicmodels Orders (5 points)


- There is a glaringly obvious design problem that could compromise data integrity in the table ```classicmodels.orders.```


- The current schema is:
```
CREATE TABLE `orders` (
  `orderNumber` int(11) NOT NULL,
  `orderDate` date NOT NULL,
  `requiredDate` date NOT NULL,
  `shippedDate` date DEFAULT NULL,
  `status` varchar(15) NOT NULL,
  `comments` text,
  `customerNumber` int(11) NOT NULL,
  PRIMARY KEY (`orderNumber`),
  KEY `customerNumber` (`customerNumber`),
  CONSTRAINT `orders_ibfk_1` FOREIGN KEY (`customerNumber`) REFERENCES `customers` (`customerNumber`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1;
```


- Alter the schema to correct the issue, and test your correction.

<u>Answer</u>

In [28]:
%%sql
ALTER TABLE
    `orders`
MODIFY COLUMN
    `status` enum(
        'Shipped',
        'Resolved',
        'Cancelled',
        'On Hold',
        'Disputed',
        'In Process'
    )
NOT NULL 

 * mysql+pymysql://dbuser:***@localhost
327 rows affected.


[]

##### Test Case: Trying to update the status to "EMBARGOED" as we did before. It should now throw an error. Testing the same below

In [29]:
try:
    %sql UPDATE orders SET status = 'EMBARGOED' WHERE customerNumber in (select customerNumber from customers where country = 'Australia') and upper(status) not in ('SHIPPED','CANCELLED')
    print("Getting here is bad.")
except Exception as e:
    print("Constraint is working, e = ", e)


 * mysql+pymysql://dbuser:***@localhost
Constraint is working, e =  (pymysql.err.DataError) (1265, "Data truncated for column 'status' at row 316")
[SQL: UPDATE orders SET status = 'EMBARGOED' WHERE customerNumber in (select customerNumber from customers where country = 'Australia') and upper(status) not in ('SHIPPED','CANCELLED')]
(Background on this error at: http://sqlalche.me/e/9h9h)


## Data Cleanup (10 Points)

- There are international standards for two letter country codes, e.g. ISO 3166-1.


- datahub.io has [down-loadable versions](https://datahub.io/core/country-list) of the information.


- The first part of answering this question is downloading the country code, country name information, and loading into a table in classicmodels.


- After a successful load, a sample query produces.

In [80]:
%sql use classicmodels;

 * mysql+pymysql://dbuser:***@localhost
0 rows affected.


[]

##### Created a table names 'countrycodes' from the downloaded data. And also added a primary key on the 'code' column which has the two digit country codes

In [72]:
%sql SELECT * FROM classicmodels.countrycodes limit 10;

 * mysql+pymysql://dbuser:***@localhost
10 rows affected.


[('Afghanistan', 'AF'),
 ('Åland Islands', 'AX'),
 ('Albania', 'AL'),
 ('Algeria', 'DZ'),
 ('American Samoa', 'AS'),
 ('Andorra', 'AD'),
 ('Angola', 'AO'),
 ('Anguilla', 'AI'),
 ('Antarctica', 'AQ'),
 ('Antigua and Barbuda', 'AG')]

##### Change the datatype on the columns, add a NOT NULL contraint and also add PRIMARY KEY to 'Code'

In [73]:
%%sql
ALTER TABLE classicmodels.countrycodes 
MODIFY COLUMN `Code` VARCHAR(2) PRIMARY KEY,
MODIFY COLUMN `Name` VARCHAR(72) NOT NULL;

 * mysql+pymysql://dbuser:***@localhost
249 rows affected.


[]

In [74]:
%sql desc classicmodels.countrycodes;

 * mysql+pymysql://dbuser:***@localhost
2 rows affected.


[('Name', 'varchar(72)', 'NO', '', None, ''),
 ('Code', 'varchar(2)', 'NO', 'PRI', None, '')]

- Allowing people to enter country names as free form text is an extremely bad idea. People will enter things like 'USA,' 'US', 'U.S. of A.', 'United States,' ...


- We are going to modify a copy of ```classicmodels.customers``` to have better integrity.


- The first step is to create a copy of ```classicmodels.customers.```

In [76]:
%sql drop table if exists classicmodels.customers_clean;

 * mysql+pymysql://dbuser:***@localhost
0 rows affected.


[]

In [77]:
%sql create table classicmodels.customers_clean as select * from classicmodels.customers;

 * mysql+pymysql://dbuser:***@localhost
122 rows affected.


[]

In [78]:
%sql select customerNumber, customerName, country from classicmodels.customers_clean limit 10;

 * mysql+pymysql://dbuser:***@localhost
10 rows affected.


[(103, 'Atelier graphique', 'France'),
 (112, 'Signal Gift Stores', 'USA'),
 (114, 'Australian Collectors, Co.', 'Australia'),
 (119, 'La Rochelle Gifts', 'France'),
 (121, 'Baane Mini Imports', 'Norway'),
 (124, 'Mini Gifts Distributors Ltd.', 'USA'),
 (125, 'Havel & Zbyszek Co', 'Poland'),
 (128, 'Blauer See Auto, Co.', 'Germany'),
 (129, 'Mini Wheels Co.', 'USA'),
 (131, 'Land of Toys Inc.', 'USA')]

##### Add a new column 'country_code' to the 'customers_clean' table. Initially it is not set to 'NOT NULL'. We'll do that later once the data is cleaned up.

In [81]:
%%sql
ALTER TABLE customers_clean 
ADD COLUMN country_code VARCHAR(2);

 * mysql+pymysql://dbuser:***@localhost
0 rows affected.


[]

##### Update the country_code column with the two character country code from the downloaded lookup table. Although it says 122 rows affected, only 80 rows have updated with the country code. We'll further clean up and update the other records

In [82]:
%%sql
UPDATE customers_clean
SET country_code = (select code from countrycodes where customers_clean.country = countrycodes.name)

 * mysql+pymysql://dbuser:***@localhost
122 rows affected.


[]

In [83]:
%sql select count(1) updated_records from customers_clean where country_code is not null

 * mysql+pymysql://dbuser:***@localhost
1 rows affected.


[(80,)]

##### Look at the countries which are not updated. This is because they are present in non-standard format.

In [84]:
%sql select distinct country from customers_clean where country_code is null

 * mysql+pymysql://dbuser:***@localhost
3 rows affected.


[('USA',), ('UK',), ('Russia',)]

##### There are 3 countries that are not in the standard format. Update the country_code for them explicitly

In [85]:
%%sql
UPDATE customers_clean SET country_code = 'US' where country = 'USA';
UPDATE customers_clean SET country_code = 'GB' where country = 'UK';
UPDATE customers_clean SET country_code = 'RU' where country = 'Russia';


 * mysql+pymysql://dbuser:***@localhost
36 rows affected.
5 rows affected.
1 rows affected.


[]

##### Now make the country_code column as NOT NULL as it now has all the records populated.

In [86]:
%%sql
ALTER TABLE customers_clean 
MODIFY COLUMN `country_code` VARCHAR(2) NOT NULL;

 * mysql+pymysql://dbuser:***@localhost
0 rows affected.


[]

##### Now let's add the PRIMARY KEY on 'customerNumber' and FOREIGN KEYS on 'salesRepEmployeeNumber' and 'country_code'

In [87]:
%%sql
ALTER TABLE customers_clean 
    ADD CONSTRAINT PRIMARY KEY (`customerNumber`),
    ADD CONSTRAINT `customers_clean_emp_number_fk_1`
          FOREIGN KEY (`salesRepEmployeeNumber`)
          REFERENCES `classicmodels`.`employees` (`employeeNumber`),
    ADD CONSTRAINT `customers_clean_ctry_code_fk_2`
          FOREIGN KEY (`country_code`)
          REFERENCES `classicmodels`.`countrycodes` (`code`);

 * mysql+pymysql://dbuser:***@localhost
122 rows affected.


[]

##### Finally let's drop the old column - 'country'

In [88]:
%%sql
ALTER TABLE customers_clean 
DROP COLUMN `country`;

 * mysql+pymysql://dbuser:***@localhost
0 rows affected.


[]

#### Describe table 'customers_clean' to look at the final changes made i.e.,  new column country_code which has the standard 2 digit codes, PRIMARY and FOREIGN Keys constraints, not null constraints etc.,

In [45]:
%%sql
desc customers_clean;

 * mysql+pymysql://dbuser:***@localhost
13 rows affected.


[('customerNumber', 'int(11)', 'NO', 'PRI', None, ''),
 ('customerName', 'varchar(50)', 'NO', '', None, ''),
 ('contactLastName', 'varchar(50)', 'NO', '', None, ''),
 ('contactFirstName', 'varchar(50)', 'NO', '', None, ''),
 ('phone', 'varchar(50)', 'NO', '', None, ''),
 ('addressLine1', 'varchar(50)', 'NO', '', None, ''),
 ('addressLine2', 'varchar(50)', 'YES', '', None, ''),
 ('city', 'varchar(50)', 'NO', '', None, ''),
 ('state', 'varchar(50)', 'YES', '', None, ''),
 ('postalCode', 'varchar(15)', 'YES', '', None, ''),
 ('salesRepEmployeeNumber', 'int(11)', 'YES', 'MUL', None, ''),
 ('creditLimit', 'decimal(10,2)', 'YES', '', None, ''),
 ('country_code', 'varchar(2)', 'NO', 'MUL', None, '')]

#####  You must produce a table that looks like the following, and implements referential integrity. Matches with the output provided

In [89]:
%sql select * from customers_clean limit 10;

 * mysql+pymysql://dbuser:***@localhost
10 rows affected.


[(103, 'Atelier graphique', 'Schmitt', 'Carine ', '40.32.2555', '54, rue Royale', None, 'Nantes', None, '44000', 1370, Decimal('21000.00'), 'FR'),
 (112, 'Signal Gift Stores', 'King', 'Jean', '7025551838', '8489 Strong St.', None, 'Las Vegas', 'NV', '83030', 1166, Decimal('71800.00'), 'US'),
 (114, 'Australian Collectors, Co.', 'Ferguson', 'Peter', '03 9520 4555', '636 St Kilda Road', 'Level 3', 'Melbourne', 'Victoria', '3004', 1611, Decimal('117300.00'), 'AU'),
 (119, 'La Rochelle Gifts', 'Labrune', 'Janine ', '40.67.8555', '67, rue des Cinquante Otages', None, 'Nantes', None, '44000', 1370, Decimal('118200.00'), 'FR'),
 (121, 'Baane Mini Imports', 'Bergulfsen', 'Jonas ', '07-98 9555', 'Erling Skakkes gate 78', None, 'Stavern', None, '4110', 1504, Decimal('81700.00'), 'NO'),
 (124, 'Mini Gifts Distributors Ltd.', 'Nelson', 'Susan', '4155551450', '5677 Strong St.', None, 'San Rafael', 'CA', '97562', 1165, Decimal('210500.00'), 'US'),
 (125, 'Havel & Zbyszek Co', 'Piestrzeniewicz', 'Zbyszek ', '(26) 642-7555', 'ul. Filtrowa 68', None, 'Warszawa', None, '01-012', None, Decimal('0.00'), 'PL'),
 (128, 'Blauer See Auto, Co.', 'Keitel', 'Roland', '+49 69 66 90 2555', 'Lyonerstr. 34', None, 'Frankfurt', None, '60528', 1504, Decimal('59700.00'), 'DE'),
 (129, 'Mini Wheels Co.', 'Murphy', 'Julie', '6505555787', '5557 North Pendale Street', None, 'San Francisco', 'CA', '94217', 1165, Decimal('64600.00'), 'US'),
 (131, 'Land of Toys Inc.', 'Lee', 'Kwai', '2125557818', '897 Long Airport Avenue', None, 'NYC', 'NY', '10022', 1323, Decimal('114900.00'), 'US')]

#### Testing of the referential integrity. Update the country code to non-standard value i.e., XX, and see exception raised for foreign key voilation.

In [47]:
try:
    %sql update customers_clean set country_code = 'XX' where customerNumber=103
    print("Getting here is bad.")
except Exception as e:
    print("This is OK, e = ", e)

 * mysql+pymysql://dbuser:***@localhost
This is OK, e =  (pymysql.err.IntegrityError) (1452, 'Cannot add or update a child row: a foreign key constraint fails (`classicmodels`.`customers_clean`, CONSTRAINT `customers_clean_ctry_code_fk_2` FOREIGN KEY (`country_code`) REFERENCES `countrycodes` (`Code`))')
[SQL: update customers_clean set country_code = 'XX' where customerNumber=103]
(Background on this error at: http://sqlalche.me/e/gkpj)


This one is really unpleasant!

<u>Answer</u>

- This one is really unpleasant.


- The character sets might be a problem.

```





```

## E-R Diagrams (5 points)

- __Note:__ Please use Crow's Foot notation for this diagram.


- The model has the following entity types:
    - Student(uni, last_name, first_name)
    - Course(course_id, course_name)
    - Section(section_number, semester, year, course_id)
    
    
- Draw a logical ER diagram representing the data model. You do not have to worry about column types.


- The model MUST represent student enrollments.


- You may need to create an additional table.

<u>Answer</u>

I've presented two approaches below on the section table. 
* In the first approach, I chose the primary key as 'callnum' (as shown in class lecture) which represents the combination of all other fields. This approach is easy to maintain. In this approach, there is a unique key on (section_number, semester, year, course_id) and a foreign key on course_id with course table. In the enrollments association table, the primary key is combination of 'uni' and 'callnum' and each of them has a foreign key to their parent tables (i.e., student and section respectively). I also made an assumption in the below that a student has to register atleast for one class and a class should have atleast one student registered.  
It is not necessary to have students in a section, but that is not the assumption here as you may see from the relationship.

<img src="./ER_diagram_surrogate_key.png">

<hr style="height:2px;">
* In the second approach, the primary key is shown as composite key on the section table. This composite key has a reference to enrollments association table i.e., all these columns has to be present in enrollments as they are all part of the key.


<img src="./ER_diagram_complex_key.png">
<hr style="height:2px;">


## Inheritance and Stored Procedures (10 points)

- The two following table definitions are a simple model for people at a university.

```
CREATE TABLE `student` (
  `student` varchar(12) NOT NULL,
  `last_name` varchar(64) NOT NULL,
  `first_name` varchar(64) NOT NULL,
  `graduation_year` year(4) NOT NULL,
  PRIMARY KEY (`student`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;


CREATE TABLE `faculty` (
  `uni` varchar(12) NOT NULL,
  `last_name` varchar(64) NOT NULL,
  `first_name` varchar(64) NOT NULL,
  `title` enum('Professor','Assistant Professor','Associate Professor','Adjunct Professor') NOT NULL,
  PRIMARY KEY (`uni`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;

```

- Implement a view ```People``` that supports ```SELECT``` for the following columns:
    - UNI
    - last_name
    - first_name
    - Type is 'S' if the person is a student and 'F' if the person is a faculty.
    - 'NA' for graduation year if the person is not a student.
    - 'NA' for title if the person is not a faculty.
    
    
- Write a stored procedure that:
    - Inserts the data in the proper table based on the type.
    - Generates a unique UNI for a newly inserted person.
    
    
- You do not need to worry about error checking parameters, types, etc.
    
    
<u>Answer</u>

```


```


In [80]:
%%sql
CREATE TABLE `student` (
  `student` varchar(12) NOT NULL,
  `last_name` varchar(64) NOT NULL,
  `first_name` varchar(64) NOT NULL,
  `graduation_year` year(4) NOT NULL,
  PRIMARY KEY (`student`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;


CREATE TABLE `faculty` (
  `uni` varchar(12) NOT NULL,
  `last_name` varchar(64) NOT NULL,
  `first_name` varchar(64) NOT NULL,
  `title` enum('Professor','Assistant Professor','Associate Professor','Adjunct Professor') NOT NULL,
  PRIMARY KEY (`uni`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;


CREATE VIEW `people` 
AS select  `student` AS `UNI`,
        `last_name` AS `last_name`,
        `first_name` AS `first_name`,
        'S' AS `type`,
        `graduation_year` AS `graduation_year`,
        'NA' AS `title`
  from `student` 
  union all 
  select `uni` AS `UNI`,
        `last_name` AS `last_name`,
        `first_name` AS `first_name`,
        'F' AS `type`,
        'NA' AS `graduation_year`,
        `title` AS `title`
  from `faculty`;


 * mysql+pymysql://dbuser:***@localhost
0 rows affected.
0 rows affected.
0 rows affected.


[]

In [52]:
%%sql
CREATE PROCEDURE `insert_data`
                            (p_fname varchar(64), 
                             p_lname varchar(64),
                             p_type varchar(1),
                             p_graduation_year year(4),
                             p_title varchar(64) 
                            )
BEGIN
    declare p_uni varchar(12);
    set p_uni = generate_uni(p_fname, p_lname);
    
    
	IF p_lname is null or p_fname is null or p_type is null then
		signal sqlstate '45002'
			set message_text="Key Fields are missing";
    END IF;
    
    
    IF p_type = 'S' then
		IF p_graduation_year is null then
			signal sqlstate '45002'
				set message_text="Graduation Year not provided";
	    END IF;
        
        INSERT INTO STUDENT (student, last_name, first_name, graduation_year)
		   VALUES (p_uni, p_lname, p_fname, p_graduation_year);  
            
    END IF;
    
    IF p_type = 'F' then
		IF p_title is null then
			signal sqlstate '45002'
				set message_text="Title not provided";
	    END IF;
        
        INSERT INTO FACULTY (uni, last_name, first_name, title)
		   VALUES (p_uni, p_lname, p_fname, p_title);  
            
    END IF;
    
    
END;


 * mysql+pymysql://dbuser:***@localhost
0 rows affected.


[]

In [53]:
%%sql
CREATE FUNCTION `generate_uni`
                    (first_name varchar(20), 
                     last_name varchar(20)) 
    RETURNS varchar(12)
    READS SQL DATA
    DETERMINISTIC
    BEGIN

    declare prefix_firstname varchar(2);
    declare prefix_lastname varchar(2);
    declare prefix varchar(4);
    declare prefix_search varchar(5);
    declare nbr_student_prefix int;
    declare nbr_faculty_prefix int;
    declare nbr_next_value int;
    declare p_uni varchar(12);
    
    set prefix_firstname = lower(substr(first_name,1,2));
    set prefix_lastname = lower(substr(last_name,1,2));
    
    set prefix = concat(prefix_firstname,prefix_lastname);
    set prefix_search = concat(prefix, '%');
    
    set nbr_student_prefix = (select count(1) as count from student where student like(prefix_search));
    set nbr_faculty_prefix = (select count(1) as count from faculty where uni like(prefix_search));
    set nbr_next_value = nbr_student_prefix + nbr_faculty_prefix + 1;
    
    set p_uni = concat(prefix, nbr_next_value);

    return p_uni;
END;

 * mysql+pymysql://dbuser:***@localhost
0 rows affected.


[]

##### Test UNI Generation function

In [54]:
%%sql
select generate_uni('Phani', 'Valasa') as uni from dual;

 * mysql+pymysql://dbuser:***@localhost
1 rows affected.


[('phva3',)]

##### Test Insertion into the Student and Faculty tables via the Procedure. Records inserted succesfully

In [56]:
%%sql

CALL insert_data('Phani', 'Valasa', 'S',  2020, null);
CALL insert_data('Phiri', 'Valasa', 'S',  2020, null);
CALL insert_data('Hank', 'Aaron', 'S',  2021, null);
CALL insert_data('Don', 'Aase', 'S',  2019, null);

CALL insert_data('Andy', 'Abad', 'F',  null, 'Professor');
CALL insert_data('Paul', 'Abbott', 'F',  null, 'Adjunct Professor');
CALL insert_data('George', 'Abrams', 'F',  null, 'Professor');
CALL insert_data('Bill', 'Abstein', 'F',  null, 'Professor');


 * mysql+pymysql://dbuser:***@localhost
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.


[]

##### Check the records inserted via the view

In [96]:
%%sql
select * From people;

 * mysql+pymysql://dbuser:***@localhost
8 rows affected.


[('doaa1', 'Aase', 'Don', 'S', '2019', 'NA'),
 ('haaa1', 'Aaron', 'Hank', 'S', '2021', 'NA'),
 ('phva1', 'Valasa', 'Phani', 'S', '2020', 'NA'),
 ('phva2', 'Valasa', 'Phiri', 'S', '2020', 'NA'),
 ('anab1', 'Abad', 'Andy', 'F', 'NA', 'Professor'),
 ('biab1', 'Abstein', 'Bill', 'F', 'NA', 'Professor'),
 ('geab1', 'Abrams', 'George', 'F', 'NA', 'Professor'),
 ('paab1', 'Abbott', 'Paul', 'F', 'NA', 'Adjunct Professor')]

##### Records from the underlying student table

In [90]:
%%sql
select * From student;

 * mysql+pymysql://dbuser:***@localhost
4 rows affected.


[('doaa1', 'Aase', 'Don', 2019),
 ('haaa1', 'Aaron', 'Hank', 2021),
 ('phva1', 'Valasa', 'Phani', 2020),
 ('phva2', 'Valasa', 'Phiri', 2020)]

##### Records from the underlying faculty table

In [91]:
%%sql
select * From faculty;

 * mysql+pymysql://dbuser:***@localhost
4 rows affected.


[('anab1', 'Abad', 'Andy', 'Professor'),
 ('biab1', 'Abstein', 'Bill', 'Professor'),
 ('geab1', 'Abrams', 'George', 'Professor'),
 ('paab1', 'Abbott', 'Paul', 'Adjunct Professor')]

##### Testing for exception when the title of the faculty is not one of 'Professor','Assistant Professor','Associate Professor' and 'Adjunct Professor'

In [93]:
try:
    %sql CALL insert_data('Andy', 'Abad', 'F',  null, 'Cafe Owner')
    print("Getting here is bad.")
except Exception as e:
    print("Exception, e = ", e)

 * mysql+pymysql://dbuser:***@localhost
Exception, e =  (pymysql.err.DataError) (1265, "Data truncated for column 'title' at row 1")
[SQL: CALL insert_data('Andy', 'Abad', 'F',  null, 'Cafe Owner')]
(Background on this error at: http://sqlalche.me/e/9h9h)


## Putting Some Pieces Together (5 points)

- The following is the current definition for ```lahman2019.salaries.```

```
CREATE TABLE `salaries` (
  `yearID` text,
  `teamID` text,
  `lgID` text,
  `playerID` text,
  `salary` text
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;

```

- Create a copy of the data into ```salaries_clean.```


- Transform the definition to improve integrity, making whatever changes you think necessary. The changes will require modifying column types, check constraints/triggers to ensure values are valid, and foreign key constraints.


<u>Answer</u>

* Change the datatype of yearID to VARCHAR(4) and make it NOT NULL. Please note this could be changed to YEAR(4) as well. However as my original schema in all tables has designed this as VARCHAR, I chose to keep it consistent.
* Change the datatype of teamID to VARCHAR(4) and make it NOT NULL
* Change the datatype of playerID to VARCHAR(16) to be consistent with other tables, and make it NOT NULL
* Change the datatpe of salary to DECIMAL(19,2) and make it NOT NULL
* lgID can be sometimes NULL, based on the distinct values seen in teams table. So, am not making this a Not NULL column. Instead am using enum() with all possible values from teams table.
* Add a primary key on (yearID, teamID and playerID). When there are multiple stints with the same team in a given year, the understanding is that salary is combined together for those stints.
* Add a foreign key on 'playerID' with 'people' table
* Add a foreign key on (teamID, yearID) with 'teams' table
* Create triggers (before insert and before update) to raise errors if the salary is Negative


```
CREATE TABLE `salaries_clean` (
  `yearID` varchar(4) NOT NULL,
  `teamID` varchar(4) NOT NULL,
  `lgID` enum('UA','AL','NL','PL','NA','AA','FL') DEFAULT NULL,
  `playerID` varchar(16) NOT NULL,
  `salary` decimal(19,2) NOT NULL,
  PRIMARY KEY (`yearID`,`teamID`,`playerID`),
  KEY `idx_team_year` (`teamID`,`yearID`),
  KEY `playerid_fk1` (`playerID`),
  CONSTRAINT `playerid_fk1` FOREIGN KEY (`playerID`) REFERENCES `people` (`playerID`),
  CONSTRAINT `team_year_fk2` FOREIGN KEY (`teamID`, `yearID`) REFERENCES `teams` (`teamID`, `yearID`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;

DROP TRIGGER IF EXISTS `lahman2019clean`.`salaries_clean_BEFORE_INSERT`;

CREATE TRIGGER `lahman2019clean`.`salaries_clean_BEFORE_INSERT` BEFORE INSERT ON `salaries_clean` FOR EACH ROW
BEGIN
    IF new.salary < 0 then
        signal sqlstate '45000'
            set message_text="Salary cannot be Negative";
    end if;
END;

DROP TRIGGER IF EXISTS `lahman2019clean`.`salaries_clean_BEFORE_UPDATE`;

CREATE TRIGGER `lahman2019clean`.`salaries_clean_BEFORE_UPDATE` BEFORE UPDATE ON `salaries_clean` FOR EACH ROW
BEGIN
    IF new.salary < 0 then
        signal sqlstate '45000'
            set message_text="Salary cannot be Negative";
    end if;
END;

```


## Graph Data $-$ Game of Thrones

- The GitHub repository https://github.com/melaniewalsh/sample-social-network-datasets/tree/master/sample-datasets/game-of-thrones contains data for a graph of relationships between characters in _Game of Thrones._
    - The file _got-nodes.csv_ contains simple information about characters.
    - The file _got-edges.csv_ contains information about relationships between characters.
    
    
- The [README](https://github.com/melaniewalsh/sample-social-network-datasets/blob/master/sample-datasets/game-of-thrones/README.md) explains the meaning of the files and fields.


- Load the data:
    - Create a new database _W4111Midterm_ in your MySQL instance.
    - Use the Table Data import tool to load the CSV files into tables named:
        - _got_nodes_
        - _got_edges_
        

- After loading, your sample data should look like the examples below.


In [97]:
%%sql
use W4111Midterm;

 * mysql+pymysql://dbuser:***@localhost
0 rows affected.


[]

In [98]:
%sql select * from W4111Midterm.got_nodes limit 10;

 * mysql+pymysql://dbuser:***@localhost
10 rows affected.


[('Aemon', 'Aemon'),
 ('Grenn', 'Grenn'),
 ('Samwell', 'Samwell'),
 ('Aerys', 'Aerys'),
 ('Jaime', 'Jaime'),
 ('Robert', 'Robert'),
 ('Tyrion', 'Tyrion'),
 ('Tywin', 'Tywin'),
 ('Alliser', 'Alliser'),
 ('Mance', 'Mance')]

In [99]:
%sql select * from W4111Midterm.got_edges limit 10;

 * mysql+pymysql://dbuser:***@localhost
10 rows affected.


[('Aemon', 'Grenn', 5),
 ('Aemon', 'Samwell', 31),
 ('Aerys', 'Jaime', 18),
 ('Aerys', 'Robert', 6),
 ('Aerys', 'Tyrion', 5),
 ('Aerys', 'Tywin', 8),
 ('Alliser', 'Mance', 5),
 ('Amory', 'Oberyn', 5),
 ('Arya', 'Anguy', 11),
 ('Arya', 'Beric', 23)]

- Wanted to know the shortest path in the data between two obscure characters: Roose, Craster.


- So, I asked my wizard friend.

| <img src="./wizard_graph.png"> |
| :---: |
| __Advanced Magic__ |


- I did not understand the spell. I have been teaching SQL. 


- But, I know giant class of wizards comfortable with SQL magic. So, I decided to ask them to show me the spell.

- So, the exam question is, "Show me an SQL spell that returns the information."


- There is a spell that is a single SQL statement, but the aliasing will drive you nuts.


- You may create tables that compute partial results.

<u>Answer</u>

##### Create a table with all direct relationships

In [100]:
%sql drop table if exists all_edges;

 * mysql+pymysql://dbuser:***@localhost
0 rows affected.


[]

In [101]:
%%sql
CREATE TABLE all_edges 
AS
SELECT Source, Target FROM got_edges
UNION
SELECT Target Source, Source Target FROM got_edges;

 * mysql+pymysql://dbuser:***@localhost
704 rows affected.


[]

##### View for relationships that are one-hop away

In [103]:
%%sql
CREATE OR REPLACE VIEW one_hop_edges AS
SELECT a.Source one_source, a.Target one_target, b.Source two_source, b.Target two_target
FROM all_edges a, all_edges b
WHERE a.Target = b.Source

 * mysql+pymysql://dbuser:***@localhost
0 rows affected.


[]

##### View for relationships that are two-hops away

In [104]:
%%sql
CREATE OR REPLACE VIEW two_hop_edges AS
SELECT a.one_source, a.one_target, a.two_source, a.two_target, b.Source one_source_1, b.Target one_target_1
FROM one_hop_edges a, all_edges b
WHERE a.two_target = b.Source

 * mysql+pymysql://dbuser:***@localhost
0 rows affected.


[]

##### Final SQL

In [105]:
%%sql
SELECT * FROM two_hop_edges
WHERE one_source in ('Craster','Roose')
AND one_target_1 in ('Craster','Roose')
AND one_source != one_target_1

 * mysql+pymysql://dbuser:***@localhost
4 rows affected.


[('Craster', 'Jon', 'Jon', 'Arya', 'Arya', 'Roose'),
 ('Roose', 'Arya', 'Arya', 'Jon', 'Jon', 'Craster'),
 ('Roose', 'Robb', 'Robb', 'Jon', 'Jon', 'Craster'),
 ('Craster', 'Jon', 'Jon', 'Robb', 'Robb', 'Roose')]

4 rows affected.


[('Craster', 'Jon', 'Jon', 'Arya', 'Arya', 'Roose'),
 ('Roose', 'Arya', 'Arya', 'Jon', 'Jon', 'Craster'),
 ('Roose', 'Robb', 'Robb', 'Jon', 'Jon', 'Craster'),
 ('Craster', 'Jon', 'Jon', 'Robb', 'Robb', 'Roose')]